### PR TITLE
Fix for communication.xml Handlers merging processs

### LINKED ISSUE #29528: [Issue] Fix for communication.xml Handlers merging processs
This issue is automatically created based on existing pull request: magento/magento2#28926: Fix for communication.xml Handlers merging processs

---------
### Description (*)
This changes fixes the "handler" merging process into communication.xml files which currently overrides all the handlers with the last one created, ignoring it's different names as expected. Issue found during custom handler development for MCOM module where there was already a handler for a message.

### Manual testing scenarios (*)
1. Develop a module which defines two handlers, with different names, for a same request on communication.xml
2. Check config after flush cache, it will only includes the last handler specified.

### Contribution checklist (*)
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] All automated tests passed successfully (all builds are green)

### PR DESCRIPTION
**Pull request by matiashidalgo**
### Description (*)
This changes fixes the "handler" merging process into communication.xml files which currently overrides all the handlers with the last one created, ignoring it's different names as expected. Issue found during custom handler development for MCOM module where there was already a handler for a message.

### Manual testing scenarios (*)
1. Please see https://github.com/magento/magento2/pull/28926#issuecomment-726669211

### Contribution checklist (*)
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] All automated tests passed successfully (all builds are green)

### Resolved issues:
1. [x] resolves magento/magento2#29528: Fix for communication.xml Handlers merging processs

### CODE REVIEW

**Comment by sidolov** [comment]
@magento create issue

**Review by lenaorobei — CHANGES_REQUESTED** [review]
Thanks for fixing it.

Could you please add integration test to cover this?

**Comment by lenaorobei** [comment]
@magento run all tests

**Comment by matiashidalgo** [comment]
> Thanks for fixing it.
> 
> Could you please add integration test to cover this?

Hi @lenaorobei since this is a fix to Framework I can not find any example of an Integration test, would you mind to point me out to an example of this and tell me which module I'm supposed to use for place my integration test?

**Comment by lenaorobei** [comment]
Sure! Please see example `dev/tests/integration/testsuite/Magento/Framework/Communication/ConfigTest.php`.

I also noticed that Static Tests build failed. Here is the report https://public-results-storage-prod.magento-testing-service.engineering/reports/magento/magento2/pull/28926/f2b73dfd3095f563cfde67e1f874d8e1/Statics/allure-report-b2b/index.html#suites/51d3ab4d5d77a0bc6a7ea344e9e7e6be/dccf717ef048c46e/

Looks like this particular case needs to be suppressed, because default parameters are different.

**Comment by lbajsarowicz** [comment]
@magento run all tests

**Comment by matiashidalgo** [comment]
@magento run all tests

**Comment by lbajsarowicz** [comment]
@matiashidalgo Perfect timing! I was just sending the recommendations to you.

**Comment by cherryz23** [comment]
Got it, thanks.

Sent from my Apple Watch

**Comment by matiashidalgo** [comment]
@lenaorobei @lbajsarowicz I can't find the issue on this failed tests, can you check?

**Comment by lenaorobei** [comment]
It might be caused by outdated branch.

@magento run all tests

**Comment by matiashidalgo** [comment]
@lenaorobei any idea why this is still failing?

**Comment by lenaorobei** [comment]
Builds fail on installation step.
```
07:00:47 [DBC DBC-SD-Tests-EE] PHP Fatal error:  Uncaught Error: Cannot instantiate interface Magento\Framework\Config\SchemaLocatorInterface in /var/www/magento2ce-upgraded/lib/internal/Magento/Framework/ObjectManager/Factory/Dynamic/Developer.php:50
07:00:47 [DBC DBC-SD-Tests-EE] Stack trace:
07:00:47 [DBC DBC-SD-Tests-EE] #0 /var/www/magento2ce-upgraded/lib/internal/Magento/Framework/ObjectManager/ObjectManager.php(70): Magento\Framework\ObjectManager\Factory\Dynamic\Developer->create()
07:00:47 [DBC DBC-SD-Tests-EE] #1 /var/www/magento2ce-upgraded/lib/internal/Magento/Framework/ObjectManager/Factory/AbstractFactory.php(170): Magento\Framework\ObjectManager\ObjectManager->get()
07:00:47 [DBC DBC-SD-Tests-EE] #2 /var/www/magento2ce-upgraded/lib/internal/Magento/Framework/ObjectManager/Factory/AbstractFactory.php(276): Magento\Framework\ObjectManager\Factory\AbstractFactory->resolveArgument()
07:00:47 [DBC DBC-SD-Tests-EE] #3 /var/www/magento2ce-upgraded/lib/internal/Magento/Framework/ObjectManager/Factory/AbstractFactory.php(239): Magento\Framework\ObjectManager\Factory\AbstractFactory->getResolvedArgument()
07:00:47 [DBC DBC-SD-Tests-EE] #4 /var/www/magento2ce-upgraded/lib/internal/Magento/Framework/ObjectManager/Factory/Dynamic/Develop in /var/www/magento2ce-upgraded/lib/internal/Magento/Framework/ObjectManager/Factory/Dynamic/Developer.php on line 50
```

**Comment by engcom-Foxtrot** [comment]
@magento run all tests

File: lib/internal/Magento/Framework/Communication/Config/Reader/XmlReader.php [hunk 1]
```diff
@@ -19,4 +23,38 @@ class XmlReader extends \Magento\Framework\Config\Reader\Filesystem
...
+    /**
+     * @param FileResolverInterface $fileResolver
+     * @param XmlReader\Converter $converter
+     * @param XmlReader\SchemaLocator $schemaLocator
+     * @param ValidationStateInterface $validationState
+     * @param string $fileName
+     * @param array $idAttributes
+     * @param string $domDocumentClass
+     * @param string $defaultScope
+     * phpcs:disable Generic.CodeAnalysis.UselessOverridingMethod
```
Comment (lbajsarowicz) [inline lib/internal/Magento/Framework/Communication/Config/Reader/XmlReader.php]: `phpcs:ignore` works on documented scope - eg. if you add `phpcs:ignore` to the class - it's being ignored in whole class, if it's applied to method - it's applied to method only.

File: lib/internal/Magento/Framework/Communication/Config/Reader/XmlReader.php [hunk 2]
```diff
@@ -19,4 +23,38 @@ class XmlReader extends \Magento\Framework\Config\Reader\Filesystem
...
+            $fileResolver,
+            $converter,
+            $schemaLocator,
+            $validationState,
+            $fileName,
+            $idAttributes,
+            $domDocumentClass,
+            $defaultScope
+        );
+    }
```
Comment (lbajsarowicz) [inline lib/internal/Magento/Framework/Communication/Config/Reader/XmlReader.php]: Whatever was hardcoded there, can't this be done with `di.xml` arguments?

**Review by lbajsarowicz — COMMENTED** [review]
@engcom-Foxtrot Please take a look at the comments.

**Comment by engcom-Foxtrot** [comment]
@magento run all tests

**Comment by engcom-Foxtrot** [comment]
@lbajsarowicz created a virtual type instead of real class.

**Comment by engcom-Foxtrot** [comment]
@magento run all tests

**Comment by lenaorobei** [comment]
@magento run all tests

**Review by lenaorobei — DISMISSED** [review]

**Comment by lbajsarowicz** [comment]
@magento run Functional Tests B2B, Functional Tests CE, Functional Tests EE

**Review by lbajsarowicz — DISMISSED** [review]
✔️  I see no more issues with the code.

**Comment by magento-engcom-team** [comment]
Hi @lbajsarowicz, thank you for the review.
[ENGCOM-8232](https://jira.corp.magento.com/browse/ENGCOM-8232) has been created to process this Pull Request

**Review by engcom-Delta — CHANGES_REQUESTED** [review]
Hi @matiashidalgo . I have tested PR changes and looks like issue is not fixed
**Manual testing scenario:**

- Add to env.php records:
```
'cron_consumers_runner' => [
        'cron_run' => true,
        'max_messages' => 20000,
        'consumers' => [
            'product_action_attribute.update',
        ]
    ],
```
- Add additional handlers for request into app/code/Magento/Catalog/etc/communication.xml
```
<topic name="product_action_attribute.update" request="Magento\AsynchronousOperations\Api\Data\OperationInterface">
        <handler name="product_action_attribute.update" type="Magento\Catalog\Model\Attribute\Backend\Consumer" method="process" />
        <handler name="product_action_attribute.update1" type="Magento\Catalog\Model\Attribute\Backend\Consumer1" method="process" />
    </topic>
```

- Add logger into handler
- - For Consumer.php:
`$this->logger->info('Original consumer runs.');`
- - For Consumer1.php:
`$this->logger->info('Second consumer runs.');`

- Go to Products grid
- Select some products using checkbox
- Choose Actios->Update Attributes
- Update Product Name and Price
- Click Save
- Run `cron:run` twice
- Open log file( in my case system.log )

**Result:**
:x: Log contains record `Original consumer runs.` only from first handler `product_action_attribute.update`. 
image

Could you take a look if steps are correct or something is missed?

**Comment by matiashidalgo** [comment]
@engcom-Delta Thanks for your test, but it is actually incorrect, when you run a consumer it took as reference queue_consumer.xml file, and it works fine if you add a line like this one:
`<consumer name="product_action_attribute.update.two" queue="product_action_attribute.update" connection="db" maxMessages="5000" consumerInstance="Magento\Framework\MessageQueue\Consumer" handler="Magento\Catalog\Model\Attribute\Backend\ConsumerTwo::process" />`
to the file **app/code/Magento/Catalog/etc/queue_consumer.xml**, despite it works it doesn't have nothing to do with communication.xml merging process, so such test will not prove that our changes fix the issue.

Please do let me know if you have any question about it.

**Comment by lbajsarowicz** [comment]
@engcom-Delta Could you take a look at the answer from @matiashidalgo ?

**Comment by engcom-Delta** [comment]
@lbajsarowicz @@matiashidalgo sorry for delay. I checked scenario according [comment](https://github.com/magento/magento2/pull/28926#issuecomment-707971913) and seems like only consumer `product_action_attribute.update.two` was run as it was included last
**Manual testing scenario:**
- Add to env.php records:
```
'cron_consumers_runner' => [
        'cron_run' => true,
        'max_messages' => 20000,
        'consumers' => [
            'product_action_attribute.update',
            'product_action_attribute.update.two',
        ]
    ],
```
- Add additional consumer into app/code/Magento/Catalog/etc/queue_consumer.xml
```
<consumer name="product_action_attribute.update.two" queue="product_action_attribute.update" connection="db" maxMessages="5000" consumerInstance="Magento\Framework\MessageQueue\Consumer" handler="Magento\Catalog\Model\Attribute\Backend\Consumer1::process" />
```

- Add logger into handler
- - For Consumer.php:
`$this->logger->info('Original consumer runs.');`
- - For Consumer1.php:
`$this->logger->info('Second consumer runs.');`

- Go to Products grid
- Select some products using checkbox
- Choose Actios->Update Attributes
- Update Product Name and Price
- Click Save
- Run `cron:run` twice
- Open log file

**Result:**
:x: Log contains record `Second consumer runs.` from last included consumer `product_action_attribute.update.two` 
#28926PRLogs

Could you provide manual testing scenario that helps verify your changes?

**Comment by engcom-Charlie** [comment]
@matiashidalgo I suppose that you fixed the next scenario:
1. Create sample `ModuleA` with `communication.xml`
```
<topic name="product_action_attribute.update" request="Magento\AsynchronousOperations\Api\Data\OperationInterface">
    <handler name="handler.moduleA" type="Magento\Catalog\Model\Attribute\Backend\Consumer" method="process" />
</topic>
```
2. Create sample `ModuleB` with `communication.xml`
```
<topic name="product_action_attribute.update" request="Magento\AsynchronousOperations\Api\Data\OperationInterface">
    <handler name="handler.moduleB" type="Magento\Catalog\Model\Attribute\Backend\Consumer" method="process" />
</topic>
```
3. Clean cache `config`
4. Make the following
```
$reader = ObjectManager::getInstance()->create(\Magento\Framework\Communication\Config\Reader\XmlReader::class);
$result = $reader->read();
```

**Before changes**
we can see only one handler for `product_action_attribute.update` topic

**After changes**
for `product_action_attribute.update` topic we can see previously created handlers

- `handler.moduleA`
- `handler.moduleB`

**Comment by matiashidalgo** [comment]
> @matiashidalgo I suppose that you fixed the next scenario:
> 
> 1. Create sample `ModuleA` with `communication.xml`
> 
> ```
> <topic name="product_action_attribute.update" request="Magento\AsynchronousOperations\Api\Data\OperationInterface">
>     <handler name="handler.moduleA" type="Magento\Catalog\Model\Attribute\Backend\Consumer" method="process" />
> </topic>
> ```
> 
> 1. Create sample `ModuleB` with `communication.xml`
> 
> ```
> <topic name="product_action_attribute.update" request="Magento\AsynchronousOperations\Api\Data\OperationInterface">
>     <handler name="handler.moduleB" type="Magento\Catalog\Model\Attribute\Backend\Consumer" method="process" />
> </topic>
> ```
> 
> 1. Clean cache `config`
> 2. Make the following
> 
> ```
> $reader = ObjectManager::getInstance()->create(\Magento\Framework\Communication\Config\Reader\XmlReader::class);
> $result = $reader->read();
> ```
> 
> **Before changes**
> we can see only one handler for `product_action_attribute.update` topic
> 
> **After changes**
> for `product_action_attribute.update` topic we can see previously created handlers
> 
> * `handler.moduleA`
> * `handler.moduleB`

Exactly, this is a correct test @engcom-Charlie

**Comment by engcom-Charlie** [comment]
@matiashidalgo could you please resolve the merge conflict?
Thank you.

**Comment by matiashidalgo** [comment]
> @matiashidalgo could you please resolve the merge conflict?
> Thank you.

Conflicts resolved! @engcom-Charlie Thank you too!

**Comment by engcom-Charlie** [comment]
@magento run all tests

**Comment by gabrieldagama** [comment]
@magento run all tests

**Comment by matiashidalgo** [comment]
@magento run Functional Tests CE

**Comment by matiashidalgo** [comment]
@engcom-Charlie @gabrieldagama all checks passed now. 

Please do let me know if something else is needed from me.

**Review by gabrieldagama — APPROVED** [review]
Hi @matiashidalgo, thanks for your contribution, the changes look good to me!

**Comment by magento-engcom-team** [comment]
Hi @gabrieldagama, thank you for the review.
[ENGCOM-8232](https://jira.corp.magento.com/browse/ENGCOM-8232) has been created to process this Pull Request

**Comment by engcom-Delta** [comment]
:heavy_check_mark: QA passed
Was checked case from Manual testing scenarios
:x: Before:
#28926Main

:heavy_check_mark: After:
#28926PR

**Comment by engcom-Delta** [comment]
**Note:** Automation tests are passed

**Pull request merged**